pkg/transfer/image/imagestore.go

```diff
@@ -163,64 +219,115 @@ func (is *Store) ImageFilter(h images.HandlerFunc, cs content.Store) images.Hand
+							Name:   fmt.Sprintf("%s@%s", ref.Name, desc.Digest),
+							Target: desc,
+							Labels: is.imageLabels,
+						})
+					}
+					continue
+				}
+
+				imgs = append(imgs, images.Image{
+					Name:   name,
+					Target: desc,
+					Labels: is.imageLabels,
+				})
+
+				// If a named reference was found and SkipNamedDigest is true, do
+				// not use this reference
+				if ref.AddDigest && !ref.SkipNamedDigest {
+					imgs = append(imgs, images.Image{
+						Name:   fmt.Sprintf("%s@%s", ref.Name, desc.Digest),
+						Target: desc,
+						Labels: is.imageLabels,
+					})
+				}
+			}
+		default:
+			return nil, fmt.Errorf("ref source not supported: %w", errdefs.ErrInvalidArgument)
 		}
-		name := imageName(desc.Annotations, nameT)
-		switch refType {
-		case "name":
-			if name == "" {
-				return images.Image{}, fmt.Errorf("no image name: %w", errdefs.ErrNotFound)
+		delete(desc.Annotations, "io.containerd.import.ref-source")
+	} else {
+		if is.imageName != "" {
+			imgs = append(imgs, images.Image{
+				Name:   is.imageName,
+				Target: desc,
+				Labels: is.imageLabels,
+			})
+		}
+
+		// If extra references, store all complete references (skip prefixes)
+		for _, ref := range is.extraReferences {
+			if ref.IsPrefix {
+				continue
 			}
-			img.Name = name
-		case "digest":
-			if !is.digestRefs || (!is.alwaysDigest && name != "") {
-				return images.Image{}, fmt.Errorf("no digest refs: %w", errdefs.ErrNotFound)
+			name := ref.Name
+			if ref.AddDigest {
+				name = fmt.Sprintf("%s@%s", name, desc.Digest)
 			}
-			img.Name = fmt.Sprintf("%s@%s", is.namePrefix, desc.Digest)
-		default:
-			return images.Image{}, fmt.Errorf("ref type not supported: %w", errdefs.ErrInvalidArgument)
+			imgs = append(imgs, images.Image{
+				Name:   name,
+				Target: desc,
+				Labels: is.imageLabels,
+			})
 		}
-		delete(desc.Annotations, "io.containerd.import.ref-type")
-	} else if img.Name == "" {
-		// No valid image combination found
-		return images.Image{}, fmt.Errorf("no image name found: %w", errdefs.ErrNotFound)
 	}
 
-	for {
-		if created, err := store.Create(ctx, img); err != nil {
+	if len(imgs) == 0 {
+		return nil, fmt.Errorf("no image name found: %w", errdefs.ErrNotFound)
+	}
+
+	for i := 0; i < len(imgs); {
+		if created, err := store.Create(ctx, imgs[i]); err != nil {
 			if !errdefs.IsAlreadyExists(err) {
-				return images.Image{}, err
+				return nil, err
 			}
 
-			updated, err := store.Update(ctx, img)
+			updated, err := store.Update(ctx, imgs[i])
 			if err != nil {
 				// if image was removed, try create again
 				if errdefs.IsNotFound(err) {
+					// Keep trying same image
 					continue
 				}
-				return images.Image{}, err
+				return nil, err
 			}
 
-			img = updated
+			imgs[i] = updated
 		} else {
-			img = created
+			imgs[i] = created
 		}
 
-		return img, nil
+		i++
 	}
+
+	return imgs, nil
 }
 
 func (is *Store) Get(ctx context.Context, store images.Store) (images.Image, error) {
```

Comment:
How could we get extra image references from the image store?

Reply:
I have that in a separate commit so far since still working on it, thinking the interface would be `Lookup(ctx context.Context, store image.Store) ([]image.Image, error)`. I was thinking how to handle the prefix references on lookup, which could be interesting to use. Non prefix references should be straight forward.